content/posts/2020/12/libplanet-0.10/index.eng.md

```diff
@@ -0,0 +1,144 @@
+[`IRenderer<T>.RenderReorg()`]: https://docs.libplanet.io/0.10.2/api/Libplanet.Blockchain.Renderers.IRenderer-1.html#Libplanet_Blockchain_Renderers_IRenderer_1_RenderReorg_Libplanet_Blocks_Block__0__Libplanet_Blocks_Block__0__Libplanet_Blocks_Block__0__
+[`LoggedRenderer<T>`]: https://docs.libplanet.io/0.10.2/api/Libplanet.Blockchain.Renderers.LoggedRenderer-1.html
+[^1]: Unlike Java, C# doesn't have an anonymous class.
+[^2]: For more information on reorg, refer to Muhun Kim's [〈Decentralization Intersects with Online Game]({{< ref "../../10/decentralized-and-online-game-intersect/index.eng.md" >}}).
+
+Delayed Renderer
+----------
+Since Libplanet currently uses a proof-of-work(PoW) method without finality, recent blocks are prone to reorg. For this reason, many cryptocurrency wallets and exchanges often show transactions' number of confirmations because the higher the number of confirmations, the less likely a reorg occurs. In order to prevent confusion in the game due to frequent reorg in the blockchain, a delayed renderer has been added. [`DelayedRenderer<T>`] is a decorator that receives `IRenderer<T>` as an input and implements `IRenderer<T>` on its own, virtually acting as a middleware that delays rendering events. So instead of immediately generating related events when new blocks are stacked in the blockchain, the delayed renderer waits until it meets the number of confirmations you have set. [Nine Chronicles] also uses a delayed renderer, and the number of confirmations is available as an option for the player to set.[^3]
+
+[`DelayedRenderer<T>`]: https://docs.libplanet.io/0.10.2/api/Libplanet.Blockchain.Renderers.DelayedRenderer-1.html
+[Nine Chronicles]: https://nine-chronicles.com/
+[^3]: The number of confirmations currently can't be adjusted on the UI, but it can be change directly in the configuration file.
+
+[Static Analyzer]
+------------------
+In Libplanet, the state within all blockchain can only be changed through an action. Because actions are executed separately on each node to derive a new state from the previous state and all nodes need to agree on a consistent state, an action must be deterministic. However, despite knowing what makes the code non-deterministic, it's not that simple to make a complex logic deterministic. You can make mistakes knowingly, and especially when multiple people work on the code, each deterministic portion implemented by different people can easily be combined to derive a non-deterministic result.
+
+To address this issue, the new version introduces [Libplanet.Analyzers Package], which checks for errors in the Libplanet action code through static analysis. This static analyzer alerts you in advance to potential bugs that are common, based on repeated mistakes we've actually made in developing Nine Chronicles. It's very easy to use, just add it as a NuGet package dependency and it will output as a warning along with a C# compiler error when built.
```

Comment:
```suggestion
To address this issue, the new version introduces [Libplanet.Analyzers Package], which checks for errors in the Libplanet action code through static analysis. This static analyzer alerts you in advance to potential bugs that are common, based on repeated mistakes we've made in developing Nine Chronicles. It's very easy to use, just add it as a NuGet package dependency and it will output as a warning along with a C# compiler error when built.
```